pkg/scheduler/api/validation/validation.go

```diff
@@ -30,14 +30,51 @@ import (
 
 // ValidatePolicy checks for errors in the Config
 // It does not return early so that it can find as many errors as possible
-func ValidatePolicy(policy schedulerapi.Policy) error {
+func ValidatePolicy(policy schedulerapi.Policy, featureDependencies []schedulerapi.FeatureDependency) error {
```

Comment:
this method should be called to see if a given config file is valid... it should generally be stateless, so it's not clear what the impact of the new second arg is. could that make a given config file sometimes valid and sometimes not?

Reply:
That's what this pull is trying to address: if a config file specifies certain policies that are technically valid but ultimately conflict based on what has been enabled or disabled through featuregates, then is that config actually valid? To answer your question, yes, this would make a given config sometimes valid and sometimes not depending on what features have been enabled for that instance of the scheduler.

Reply:
@liggitt bump, is that sufficient? Would you rather see any changes?

Reply:
one last question... I'm trying to understand the order and dependency between scheduler config (via flags or config file), scheduler policy config (via file), and how the feature flags plumbing affects our ability to evaluate if a given config is valid before making it effective.

1. feature flags come from CLI flags (today), but as scheduler config file work progresses, I'd expect them to be able to be specified in the config file
2. the scheduler config file can reference a policy file
3. part of validating the scheduler config would be to load and validate the policy as well

We want to be able to validate config file and policy file before making them effective. That means all validation functions (and the things that compute the inputs to those functions) must be able to be evaluated without affecting the global effective state.

The primary place I see that needs changing is we have to be able to compute the inputs to validation without changing global state. That means we need a function that computes what ApplyFeatureGates() is returning that takes a featuregate as an argument (doesn't use the global `utilfeature.DefaultFeatureGate`) and returns computed values without modifying the global factory. That way, to validate a config prior to making it effective, we would:
1. load the scheduler config file
2. apply the feature gates specified in the config file to `temporaryFeatureGateSet`
3. call `ComputeFeatureGates(temporaryFeatureGateSet)` (which would have no side effects)
4. validate the policy config passing in the computed feature gates

Reply:
I agree with @liggitt that we shouldn't calculate feature dependencies in ApplyFeatureGates() - see https://github.com/kubernetes/kubernetes/pull/79426/files#r302785224

---

More thoughts on the underlying configuration issue:

Essentially how default scheduler behaves depends on (a) feature gates plus (b) predicates/priorities, the combinations could be:

1. feature gates + default policy (i.e. default predicates/priorities)
1. feature gates + custom policy (i.e. custom predicates/priorities)

Let's focus on the 2nd case here. The following cases we should give validation error:

- custom policy violates (non-GA) feature gates
	This can break the function of the whole kubernetes cluster. E.g. an alpha feature functions across apiserver and scheduler, if the related predicate and/or priority is specified in custom policy file, it may work as the feature gate is disabled in apiserver side. (NOTE: we don't think it's a violation if a feature gate is enabled, but its predicate/priority is missing)
- custom policy violates predicate/priority dependencies
	For example, a predicate is specified in policy file, but its dependent predicate/priority isn't specified, or its violated predicate/priority is specified.
- custom policy violates generic mandatory predicate/priority
	That is to say, there may be some generic mandatory predicate/priority, if it's missing (from custom policy file), scheduler may not work. Right now I didn't see any kind of pred/prio falls into this category, but I may be wrong, or in the future we may have some.

Based on the above conditions, I'd suppose:

1. Maintain a data structure to have mappings from:
	1. feature gate => predicate/priority
	1. predicate/priority => dependent predicate/priority and violated predicate/priority
1. The mapping structure should be static (for a single release), so put it somewhere as a simple `var`
1. Per `scheduler config validation`, do NOT run `ApplyFeatureGates()`; instead get the mapping obj, and ensure all violations are captured.
1. (I will propose another PR) Based on the mapping structure, revise ApplyFeatureGates() - this function looks buggy, esp. when used in integration tests.

@damemi, thoughts?

Reply:
And (in another PR), we should discuss whether we should or should not apply `ApplyFeatureGates()` upon a custom policy config file. If we did, it seems we **mutates** the predicates/priorities collection given by user.